hail/python/hail/table.py

```diff
@@ -2310,17 +2310,16 @@ def join(self,
 
         Notes
         -----
-        Hail supports four types of joins specified by `how`:
-
-        - **inner** -- Key must be present in both the left and right tables.
-        - **outer** -- Key present in either the left or the right. For keys
-          only in the left table, the right table's fields will be missing.
-          For keys only in the right table, the left table's fields will be
-          missing.
-        - **left** -- Key present in the left table. For keys not found on
-          the right, the right table's fields will be missing.
-        - **right** -- Key present in the right table. For keys not found on
-          the right, the right table's fields will be missing.
+        Hail supports four types of joins specified by `how`.
+
+        - **inner** -- Key values must match in both the left and right tables for a record to be in the joined table.
```

Comment:
the phase "key value" is probably going to confuse people. We should maybe also mention the fact that we do the cartesian product of matching keys.

Reply:
That is mentioned below

Reply:
https://github.com/hail-is/hail/blob/7eeb6c122a761e659d505ac834dbb34b992f3ba3/hail/python/hail/table.py#L2340

Reply:
> key value

It will be more confusing to confuse “keys” and “values of key column” as we do now, especially since above this we speak about the necessities for keys to be precisely matched in type, order, and number. 

Here is an example of how this is described in the relational space:

“ Then, we can create the following SQL statement (that contains an INNER JOIN), that selects records that have matching values in both tables:”

https://www.w3schools.com/sql/sql_join.asp

I think this is clear. What about an updated introductory statement:“Tables are joined by comparing the values of their keys, row by row. How this comparison is made depends on the join strategy:” And then continue with the explanation for each join type. Sound reasonable?

Reply:
ah, missed that we have the cartesian product doc below, that seems fine.

Reply:
yeah, I like an introduction. Slight tweak on yours: "Tables are joined at rows that have the same value of non-missing key fields. The inclusion of rows with no match depends on the join strategy:"